[WebGPU] GPURenderBundleEncoder.drawIndirect and GPURenderBundleEncoder.drawIndexedIndirect have synchronization issues

mwyrzykowski opened the pull request:
#### f915246e31c3594167be28c244e9790100234ecc
<pre>
[WebGPU] GPURenderBundleEncoder.drawIndirect and GPURenderBundleEncoder.drawIndexedIndirect have synchronization issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=264213">https://bugs.webkit.org/show_bug.cgi?id=264213</a>
&lt;radar://117948947&gt;

Reviewed by Dan Glastonbury.

RenderBundleEncoder supports drawIndirect and drawIndexedIndirect but
MTLIndirectRenderCommand does not. Initially this was implemented by directly
reading the contents property of the indirect buffer, however that has synchronization
issues if the buffer is encoded in a copyBufferToBuffer command or similar.

Eventually we would like to build the ICBs via a compute kernel, as noted in
<a href="https://bugs.webkit.org/show_bug.cgi?id=264219.">https://bugs.webkit.org/show_bug.cgi?id=264219.</a>

However, until <a href="https://bugs.webkit.org/show_bug.cgi?id=264219">https://bugs.webkit.org/show_bug.cgi?id=264219</a> is implemented we will
support RenderBundles via command replay to avoid the synchronization issues described.

This also allows all the memory_sync tests to pass, so enable them.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/memory_sync/buffer/multiple_buffers-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/memory_sync/buffer/single_buffer-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/memory_sync/texture/same_subresource-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::addResource):
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):
(WebGPU::RenderBundleEncoder::drawIndirect):

* LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/programmable/state_tracking-expected.txt:

Canonical link: <a href="https://commits.webkit.org/270311@main">https://commits.webkit.org/270311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5a314efdfd335c4b4ec323c4c141b15921d3203

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25073 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26334 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27191 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23018 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5319 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1055 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23280 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25316 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2654 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21645 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27772 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2348 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22581 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28712 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22883 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22937 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26583 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2287 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/595 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3620 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6021 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2733 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2630 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WebGPU/RenderBundleEncoder.h:
```diff
@@ -140,7 +140,8 @@ class RenderBundleEncoder : public WGPURenderBundleEncoderImpl, public RefCounte
     const RenderPipeline* m_pipeline { nullptr };
-    HashMap<uint32_t, Vector<uint32_t>, DefaultHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_bindGroupDynamicOffsets;
+    using BindGroupDynamicOffsetsContainer = HashMap<uint32_t, Vector<uint32_t>, DefaultHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>>;
+    std::optional<BindGroupDynamicOffsetsContainer> m_bindGroupDynamicOffsets;
```
Is an empty `BindGroupDynamicOffsetsContainer` equivalent to the "None" case of `std::optional`? (ie. I don't think the `std::optional` is required)

mwyrzykowski replied:
I'm using std::optional to avoid building the dynamic offsets vertex and fragment buffers more than once.

`RenderBundleEncoder::replayCommands` sets `m_bindGroupDynamicOffsets = std::nullopt;` such that `m_dynamicOffsetsVertexBuffer` and `m_dynamicOffsetsFragmentBuffer` are not continuously rebuilt in `RenderBundleEncoder::executePreDrawCommands` since the contents can't change.

I understand this might be confusing as typing up that explanation was tricky to word correctly, so I am open to alternatives to achieving this

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU] GPURenderBundleEncoder.drawIndirect and GPURenderBundleEncoder.drawIndexedIndirect have synchronization issues
https://bugs.webkit.org/show_bug.cgi?id=264213
<radar://117948947>

Reviewed by Dan Glastonbury.

RenderBundleEncoder supports drawIndirect and drawIndexedIndirect but
MTLIndirectRenderCommand does not. Initially this was implemented by directly
reading the contents property of the indirect buffer, however that has synchronization
issues if the buffer is encoded in a copyBufferToBuffer command or similar.

Eventually we would like to build the ICBs via a compute kernel, as noted in
https://bugs.webkit.org/show_bug.cgi?id=264219.

However, until https://bugs.webkit.org/show_bug.cgi?id=264219 is implemented we will
support RenderBundles via command replay to avoid the synchronization issues described.

This also allows all the memory_sync tests to pass, so enable them.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/memory_sync/buffer/multiple_buffers-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/memory_sync/buffer/single_buffer-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/memory_sync/texture/same_subresource-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::addResource):
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):
(WebGPU::RenderBundleEncoder::drawIndirect):

* LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/programmable/state_tracking-expected.txt:

Canonical link: https://commits.webkit.org/270311@main
